driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+        }
+
+        ServerSelector apply(final ServerSelector selector) {
+            return new DeprioritizingSelector(selector);
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
```

Comment:
We should put all these conditions into the same place, `onAttemptFailure`.

Alternatively, we could skip adding a DeprioritizingSelector based on this check. (Whatever we do, I would expect disabling deprioritization in this way to just work.)

Reply:
When we discussed this, we found that `binding.getWriteConnectionSource().getServerDescription().getClusterType()` can be used to trivially get `ClusterType` that can be passed into `ServerDeprioritization.onAttemptFailure`. However, after looking a bit deeper, I noticed that `binding.getWriteConnectionSource()` actually does server selection, so we really can't call this method when we create `BinaryOperator<Throwable> onAttemptFailure` in `decorateWriteWithRetries`.

We still may attach `ClusterType` from `binding to `RetryState` when we currently call `binding.getWriteConnectionSource()`, and then access it from the `BinaryOperator<Throwable> onAttemptFailure` object via the `RetryState`, but then the change becomes not trivial at all.

Thus, I don't see a trivial way to pass `ClusterType` into `ServerDeprioritization.onAttemptFailure`. The solution I see is not worth the potential simplification of moving the `isEnabled(clusterType)` check from `ServerDeprioritization.updateCandidate` to `ServerDeprioritization.onAttemptFailure`.